dim/src/external/tmdb/metadata_provider.rs

```diff
@@ -0,0 +1,298 @@
+
+        let api_key: Arc<str> = api_key.to_owned().into_boxed_str().into();
+
+        Self {
+            api_key,
+            http_client,
+            cache: Default::default(),
+        }
+    }
+
+    /// curry this metadata provider to supply search results for TV shows.
+    #[inline(always)]
+    pub fn tv_shows(&self) -> MetadataProviderOf<TvShows> {
+        MetadataProviderOf {
+            provider: self.clone(),
+            _key: PhantomData,
+        }
+    }
+
+    #[inline(always)]
+    pub fn movies(&self) -> MetadataProviderOf<Movies> {
+        MetadataProviderOf {
+            provider: self.clone(),
+            _key: PhantomData,
+        }
+    }
+
+    /// insert a default [CacheValue] if the slot at a given key is not present.
+    fn insert_value_if_empty(&self, key: &CacheKey) -> (CacheValue, bool) {
+        // grab the entry or instert RwLock::new(None) if not present.
+        let mut entry = self.cache.entry(key.clone()).or_default();
+
+        // fast path: cache hits, no writers and the value is present.
+        {
+            let read_guard = entry.value();
+            if let Some(value) = read_guard.as_ref() {
+                return (value.clone(), false);
+            }
+        }
+
+        // slow path: get a write guard, if the slot is still uninit when we acquire; initialize it.s
+        let slot = entry.value_mut();
+
+        match slot.as_ref() {
+            // someone initialized the slot before we got the write guard, use their value.
+            Some(value) => (value.clone(), false),
+            // we're still first, initialize the value and keep going.
+            None => {
+                let (tx, _) = broadcast::channel(1);
+                let value = CacheValue::RequestInFlight { tx };
+                slot.replace(value.clone());
+                (value, true)
+            }
+        }
+    }
+
+    /// perform request coalescing; when two futures are made with the same key the duplicates wait for the original to broadcast the results.
+    async fn coalesce_request<F, Fut>(
+        &self,
+        key: &CacheKey,
+        make_request_future: F,
+    ) -> Result<Arc<str>, Error>
+    where
+        F: FnOnce(TMDBClient) -> Fut,
+        Fut: Future<Output = Result<Arc<str>, TMDBClientRequestError>> + Send + 'static,
+    {
+        let (value, we_own_future) = { self.insert_value_if_empty(key) };
+
+        match (we_own_future, value) {
+            // only if we own the future that was spawned should we upate the
+            // value once we get the response.
+            (true, CacheValue::RequestInFlight { tx }) => {
+                let client = TMDBClient {
+                    provider: self.clone(),
+                };
+
+                let tx_ = tx.clone();
+                let fut = make_request_future(client);
+                tokio::spawn(async move {
+                    let output = fut.await;
+                    let _ = tx_.send(output);
+                });
```

Comment:
wonder if this could just be a ```tokio::spawm(fut.then(|x| tx_.send(x)))```

Reply:
Probably not. I think this modifies the return type to not be `()`